### PR TITLE
Feat: Add Support for Linux Users and Custom Install Script

### LINKED ISSUE #594: Files Written to Mounted Linux Home Directory Owned by Root User
## what
* The user's shell inside Geodesic runs as `root`
* The script that launches Geodesic bind-mounts the host user's `$HOME` to `/localhost` to provide access to configuration files and allow for editing of host files
* Depending on the storage driver being used, it is possible that files created under `/localhost` from within Geodesic will be set to the same owner UID and GID (that is, owned by `root`) on the host as they have within Geodesic.
* Notably, this is _**not an issue**_ with Docker Desktop for Mac, which uses the `osxfs` storage driver to maintain a distinction between the UID and GID of a file as seen from within the Docker container versus what is set on the host file system. See [Ownership](https://web.archive.org/web/20190719101305/https://docs.docker.com/docker-for-mac/osxfs/#ownership) section of documentation.
* Update Feb 2022: this is _**not an issue**_ with Docker v20 on Ubuntu (tested on 20.04 LTS) when running in [rootless mode](https://docs.docker.com/engine/security/rootless/). As with macOS, this configuration correctly translates file ownership between the root UID and GID inside the container to the user's UID and GID on the host. 

## remediation discussion

Under MacOS, this is not a problem because the [OSXFS](https://docs.docker.com/docker-for-mac/osxfs/#ownership) driver handles UID and GID remapping automatically.

Under Linux, it is possible to use [User Namespaces](https://docs.docker.com/engine/security/userns-remap/) to remap the root user UID 0 to another UID (presumably the UID of the user running Geodesic), and do the same for GID, however, the documentation says you cannot use remapping when using the `--privileged` flag, which Geodesic requires, and even if you could, it would cause problems with `sudo` and setuid binaries in the container. 

#### further reading

- https://docs.docker.com/engine/security/userns-remap/
- https://blog.gougousis.net/file-permissions-the-painful-side-of-docker/
- https://seravo.fi/2019/align-user-ids-inside-and-outside-docker-with-subuser-mapping

### PR DESCRIPTION
**Pull request by drmikecrowe**
* Support Linux non-root user -- fixes #594
* Add customization instructions to README
* Allow mapping of volumes in installation script

### CODE REVIEW

**Comment by korenyoni** [comment]
Hi @drmikecrowe, thanks for your first contribution!

Next time, please make use of `PULL_REQUEST_TEMPLATE`, i.e. preserving the _what_ and _why_ sections.

In terms of requesting changes or approval, I'm going to defer this to @Nuru

File: README.yaml [hunk 1]
```diff
@@ -158,7 +158,40 @@ usage: |-
   Detailed information about launch-time configuration is in the [customization](./docs/customization.md)
   document, available from within the shell via `man customization`.
 
+  #### Customizing the installation script
```
Comment (korenyoni) [inline README.yaml:161]: Should this be in https://github.com/cloudposse/geodesic/blob/8f987c7b3e7ec67449c886dd2a7412bb52d7e796/docs/customization.md instead?

File: rootfs/templates/wrapper [hunk 2]
```diff
@@ -187,7 +194,7 @@ function use() {
 		echo "# Starting new ${DOCKER_NAME} session from ${DOCKER_IMAGE}"
 		echo "# Exposing port ${GEODESIC_PORT}"
 		[ -z "${GEODESIC_DOCKER_EXTRA_ARGS}" ] || echo "# Launching with extra Docker args: ${GEODESIC_DOCKER_EXTRA_ARGS}"
-		docker run "${DOCKER_ARGS[@]}" ${GEODESIC_DOCKER_EXTRA_ARGS} ${DOCKER_IMAGE} -l $*
+		docker run "${DOCKER_ARGS[@]}" ${GEODESIC_DOCKER_EXTRA_ARGS} ${GEODESIC_ADDITIONAL_DOCKER_ARGS} ${DOCKER_IMAGE} -l $*
```
Comment (korenyoni) [inline rootfs/templates/wrapper:197]: In terms of naming, it's hard to distinguish `GEODESIC_DOCKER_EXTRA_ARGS` and `GEODESIC_ADDITIONAL_DOCKER_ARGS`, they mean the same thing

File: rootfs/templates/wrapper [hunk 3]
```diff
@@ -16,6 +16,8 @@ export GROUP_ID=$(id -g)
 export options=()
 export targets=()
 
+export GEODESIC_ADDITIONAL_DOCKER_ARGS="{{getenv "GEODESIC_ADDITIONAL_DOCKER_ARGS" ""}}"
```
Comment (korenyoni) [inline rootfs/templates/wrapper:19]: See note below on naming

**Review by korenyoni — COMMENTED** [review]
Some comments.

Again, I want to defer to @Nuru because he will likely suggest a different high level strategy for achieving this functionality.

**Comment by drmikecrowe** [comment]
> Next time, please make use of `PULL_REQUEST_TEMPLATE`, i.e. preserving the _what_ and _why_ sections.
Let me resubmit, and pull apart the customization vs. fixing #594

**Pull request closed without merging**